### PR TITLE
fix: Adjust check of historical Project identifiers upon saving

### PR DESCRIPTION
**Pull request by thykel**
# Ticket

# What are you trying to accomplish?
Migrating the edge QA instance from `classic` to `semantic` and back to `classic` revealed an edge case:

When database contains the two following projects: 
* Project 1, `identifier: 'TEST'`, `slugs: ['TEST, 'testrobin812345']`
* Project 2, `identifier: 'SP2'`, `slugs: ['SP2', 'test']`

...then the process gets stuck at rename of the second project, since renaming to `test` fails the `identifier_not_historically_reserved` validation due to conflict with `TEST` (even thought at that point, that's no longer the current identifier).

# What approach did you choose and why?
The easy solution is to relax the `slugs` lookup from case-insensitive to case-sensitive. The sets of Ids from each of the two modes shouldn't affect each other in terms of uniqueness.

# Merge checklist

- [ ] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

### CODE REVIEW

File: app/models/projects/identifier.rb [hunk 1]
```diff
@@ -162,7 +162,7 @@ def identifier_not_historically_reserved
     return if errors.any? { |error| error.attribute == :identifier && error.type == :taken }
 
     already_existing = FriendlyId::Slug
-                         .where("LOWER(slug) = LOWER(?)", identifier)
+                         .where(slug: identifier)
```
Comment (akabiru) [inline app/models/projects/identifier.rb:165]: > When database contains the two following projects:
> 
>     * Project 1, identifier: 'TEST', slugs: ['TEST, 'testrobin812345']
>     * Project 2, identifier: 'SP2', slugs: ['SP2', 'test']
> 
> The easy solution is to relax the slugs lookup from case-insensitive to case-sensitive. The sets of Ids from each of the two modes shouldn't affect each other in terms of uniqueness.

This would mean that we'll be allowing the two projects to be historically accessible via the same identifier with just different casing? That sounds super brittle to me. This is the primary reason we implemented case-insensitive uniqueness.

Other options that come to mind:

* Deduplication via FIFO: Project 1 retains the “Test” identifier, while Project 2 is renamed to “test1”. This is the same collision algorithm employed when deduplicating identifier suggestions.
* Deleting other historical collissions, keeping only the first. Project 1 gets to keep "TEST", Project 2 does not.

https://github.com/opf/openproject/blob/d309f2a419677cacb9f12ac655be0555a34a3bbc/app/services/project_identifiers/identifier_autofix/project_identifier_suggestion_generator.rb#L198-L213

**Review by akabiru — CHANGES_REQUESTED** [review]

File: app/models/projects/identifier.rb (hunk 1 above)
Comment (thykel) [reply app/models/projects/identifier.rb:165]: > This would mean that we'll be allowing the two projects to be historically accessible via the same identifier with just different casing? That sounds super brittle to me. This is the primary reason we implemented case-insensitive uniqueness.

OK, fair point. Then we have to shift focus to the convertor to `semantic` mode, because _that_ was the one that caused the DB to land in that state. Checking.

Comment (thykel) [reply app/models/projects/identifier.rb:165]: Ah, yes, this is caused by the [skipped `FriendlyId` uniqueness validation](https://github.com/opf/openproject/blob/9b38c082ab684ee37e20613a045ce12bec06d168/app/services/project_identifiers/convert_project_to_semantic_service.rb#L87-L90). Will see if I can can make that step smarter.

Comment (thykel) [reply app/models/projects/identifier.rb:165]: @judithroth actually already predicted this [here](https://github.com/opf/openproject/pull/22765#discussion_r3111337112). I _did_ fix these gaps for (re-)generation of _classic_ identifiers, but the issue is still there for conversion to `semantic`. We have to adjust the semantic `ProjectIdentifierSuggestionGenerator` to avoid reserved keyword and historical aliases.

(And then preferably make sure the AR validations actually run.)

Comment (judithroth) [reply app/models/projects/identifier.rb:165]: Like I said in https://github.com/opf/openproject/pull/22765#discussion_r3111337112 omitting validations on the attribute we want to change will lead to problems. (I am open to discuss omitting validations on other attributes though).

Comment (judithroth) [reply app/models/projects/identifier.rb:165]: > [@judithroth](https://github.com/judithroth) actually already predicted this [here](https://github.com/opf/openproject/pull/22765#discussion_r3111337112). I _did_ fix these gaps for (re-)generation of _classic_ identifiers, but the issue is still there for conversion to `semantic`. We have to adjust the semantic `ProjectIdentifierSuggestionGenerator` to avoid reserved keyword and historical aliases.
> 
> (And then preferably make sure the AR validations actually run.)

Thanks Tom 👍

**Comment by thykel** [comment]
Superseded by https://github.com/opf/openproject/pull/22932
